### PR TITLE
Noita: Update to use new Options API

### PR DESCRIPTION
**Pull request by ScipioWright**
## What is this fixing or adding?
Reworking the options to make it work with the new options API.
Also reworked stuff in several spots to use world: NoitaWorld instead of multiworld: MultiWorld

## How was this tested?
Generated a few games with different settings, connected it to the mod once

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

File: worlds/noita/__init__.py [hunk 1]
```diff
@@ -36,21 +38,20 @@ class NoitaWorld(World):
     web = NoitaWeb()
 
     # Returned items will be sent over to the client
-    def fill_slot_data(self):
-        return {name: getattr(self.multiworld, name)[self.player].value for name in self.option_definitions}
+    # def fill_slot_data(self):
```
Comment (heinermann) [inline worlds/noita/__init__.py:41]: What's up with this?

Comment (ScipioWright) [reply worlds/noita/__init__.py:41]: A mistake, woops

File: worlds/noita/__init__.py [hunk 2]
```diff
@@ -36,21 +39,32 @@ class NoitaWorld(World):
     web = NoitaWeb()
 
     # Returned items will be sent over to the client
-    def fill_slot_data(self):
-        return {name: getattr(self.multiworld, name)[self.player].value for name in self.option_definitions}
+    def fill_slot_data(self) -> Dict[str, Any]:
+        return {
```
Comment (alwaysintreble) [inline worlds/noita/__init__.py:43]:
```suggestion
        return self.options.as_dict("death_link", "bad_effects", etc)
```
do you really need the start inventory in your slot data? please don't send data you don't need.

Comment (ScipioWright) [reply worlds/noita/__init__.py]: Removed that and bad effects, forgot that start_inventory_from_pool was handled with item sends rather than something the client needs, been a long day

File: worlds/noita/Rules.py [hunk 3]
```diff
@@ -66,9 +71,9 @@ def has_orb_count(state: CollectionState, player: int, amount: int) -> bool:
     return state.item_count("Orb", player) >= amount
 
 
-def forbid_items_at_location(multiworld: MultiWorld, location_name: str, items: Set[str], player: int):
-    location = multiworld.get_location(location_name, player)
-    GenericRules.forbid_items_for_player(location, items, player)
+def forbid_items_at_location(world: NoitaWorld, location_name: str, items: Optional[Set[str] | List[str]]):
```
[the remark this answers is not in the document]
Comment (ScipioWright) [reply worlds/noita/Rules.py:74]: Unfortunately, `forbid_items_for_player` takes a `Set[str]`, so we'd have to instead convert everything into sets. I'll leave it up to you, I don't really mind either way.

[the remark this answers is not in the document]
Comment (heinermann) [reply worlds/noita/Rules.py:74]: In that case, `items_hidden_from_shops` can be made into a set naturally, and other stuff can just use the set constructor `wands_to_forbid = set(wand_tiers[whatever])`

**Comment by heinermann** [comment]
You should be able to delete the `add_location` function in `Regions.py`.

File: worlds/noita/Events.py [hunk 4]
```diff
@@ -12,21 +11,20 @@ def create_location(player: int, name: str, region: Region) -> Location:
...
     new_location.place_locked_item(create_event(player, item))
 
     region.locations.append(new_location)
     return new_location
 
 
-def create_all_events(multiworld: MultiWorld, player: int) -> None:
-    for region, event in event_locks.items():
-        create_locked_location_event(multiworld, player, region, event)
+def create_all_events(world, created_regions: Dict[str, Region]) -> None:
```
Comment (alwaysintreble) [inline worlds/noita/Events.py:22]: why do you give typing to world everywhere else but here?

File: worlds/noita/Items.py [hunk 5]
```diff
@@ -44,32 +49,32 @@ def create_kantele(victory_condition: VictoryCondition) -> List[str]:
     return ["Kantele"] if victory_condition.value >= VictoryCondition.option_pure_ending else []
 
 
-def create_random_items(multiworld: MultiWorld, player: int, random_count: int) -> List[str]:
+def create_random_items(world: NoitaWorld, random_count: int) -> List[str]:
     filler_pool = filler_weights.copy()
-    if multiworld.bad_effects[player].value == 0:
+    if world.options.bad_effects.value == 0:
```
Comment (alwaysintreble) [inline worlds/noita/Items.py:54]:
```suggestion
    if not world.options.bad_effects:
```

File: worlds/noita/Regions.py [hunk 6]
```diff
@@ -1,48 +1,51 @@
...
     for location_name, location_data in locations.items():
         location_type = location_data.ltype
         flag = location_data.flag
 
-        opt_orbs = multiworld.orbs_as_checks[player].value
-        opt_bosses = multiworld.bosses_as_checks[player].value
-        opt_paths = multiworld.path_option[player].value
-        opt_num_chests = multiworld.hidden_chests[player].value
-        opt_num_pedestals = multiworld.pedestal_checks[player].value
+        opt_orbs = world.options.orbs_as_checks.value
```
Comment (alwaysintreble) [inline worlds/noita/Regions.py:20]: i really think this reads better without these extra variable definitions. i also don't see anywhere that the `.value` is actually necessary.

File: worlds/noita/Regions.py [hunk 7]
```diff
@@ -1,48 +1,51 @@
...
         elif location_type == "chest" and flag <= opt_paths:
-            for i in range(opt_num_chests):
-                add_location(player, f"{location_name} {i+1}", location_data.id + i, region)
+            amount = opt_num_chests
         elif location_type == "pedestal" and flag <= opt_paths:
-            for i in range(opt_num_pedestals):
-                add_location(player, f"{location_name} {i+1}", location_data.id + i, region)
+            amount = opt_num_pedestals
+
+        region.add_locations(Locations.make_location_range(location_name, location_data.id, amount),
```
Comment (alwaysintreble) [inline worlds/noita/Regions.py:36]: other than here ig

File: worlds/noita/Regions.py [hunk 8]
```diff
@@ -1,48 +1,51 @@
...
+                             Locations.NoitaLocation)
 
 
 # Creates a new Region with the locations found in `location_region_mapping` and adds them to the world.
-def create_region(multiworld: MultiWorld, player: int, region_name: str) -> Region:
-    new_region = Region(region_name, player, multiworld)
-    add_locations(multiworld, player, new_region)
+def create_region(world: NoitaWorld, region_name: str) -> Region:
+    new_region = Region(region_name, world.player, world.multiworld)
+    add_locations(world, new_region)
```
Comment (alwaysintreble) [inline worlds/noita/Regions.py:43]: i really don't think it's good practice to have multiple nested function calls that use the exact same name.

File: worlds/noita/Regions.py [hunk 9]
```diff
@@ -1,48 +1,51 @@
...
     return new_region
 
 
-def create_regions(multiworld: MultiWorld, player: int) -> Dict[str, Region]:
-    return {name: create_region(multiworld, player, name) for name in noita_regions}
+def create_regions(world: NoitaWorld) -> Dict[str, Region]:
```
Comment (alwaysintreble) [inline worlds/noita/Regions.py:47]: why does this function even exist?

Comment (alwaysintreble) [reply worlds/noita/Regions.py:47]: can't mark it, but same comment for create_item in Items.py line 25. that could just be the definition in the world and then all you have to do is `world.create_item(item_name)`

File: worlds/noita/Rules.py [hunk 10]
```diff
@@ -66,9 +71,9 @@ def has_orb_count(state: CollectionState, player: int, amount: int) -> bool:
     return state.item_count("Orb", player) >= amount
 
 
-def forbid_items_at_location(multiworld: MultiWorld, location_name: str, items: Set[str], player: int):
-    location = multiworld.get_location(location_name, player)
-    GenericRules.forbid_items_for_player(location, items, player)
+def forbid_items_at_location(world: NoitaWorld, location_name: str, items: Set[str]):
```
Comment (alwaysintreble) [inline worlds/noita/Rules.py:74]: I think it'd make more sense to either just directly pass the location to this method or remove this function all together personally, but this one isn't that big of a deal. I understand the idea of wanting to segment everything apart, but you have everything so segmented that it makes it more difficult to follow what's actually happening.

File: worlds/noita/__init__.py [hunk 11]
```diff
@@ -36,21 +39,21 @@ class NoitaWorld(World):
...
 
     def create_regions(self) -> None:
-        Regions.create_all_regions_and_connections(self.multiworld, self.player)
-        Events.create_all_events(self.multiworld, self.player)
+        Regions.create_all_regions_and_connections(self)
 
     def create_item(self, name: str) -> Item:
         return Items.create_item(self.player, name)
 
     def create_items(self) -> None:
```
Comment (alwaysintreble) [inline worlds/noita/__init__.py:52]: I wouldn't deny based on this, but I really don't like all these nested calls just for the sake of it. it's hard to follow, and every function created takes up more memory and time.

**Review by alwaysintreble — CHANGES_REQUESTED** [review]

File: worlds/noita/Regions.py (hunk 8 above)
Comment (heinermann) [reply worlds/noita/Regions.py:43]: `add_locations` can be renamed to `create_locations`.

File: worlds/noita/Regions.py (hunk 9 above)
Comment (heinermann) [reply worlds/noita/Regions.py:47]: > why does this function even exist?

To make `create_all_regions_and_connections` readable and keep the world class slim.

File: worlds/noita/Rules.py (hunk 10 above)
Comment (heinermann) [reply worlds/noita/Rules.py:74]: The location needs to be retrieved in every case which is what this is really facilitating. If anything more logic could be moved into this function to handle multiple locations, then code at three of the callers can be improved such as.

```py
def ban_items_from_shops(multiworld: MultiWorld, player: int) -> None:
    for location_name in Locations.location_name_to_id.keys():
        if "Shop Item" in location_name:
            forbid_items_at_location(multiworld, location_name, items_hidden_from_shops, player)
```

could be

```py
def ban_items_from_shops(multiworld: MultiWorld, player: int) -> None:
    shop_locations = [name for name in Locations.location_name_to_id.keys() if "Shop Item" in name]
    forbid_items_at_locations(multiworld, shop_locations, items_hidden_from_shops, player)
```

The two for loops for the other callers would be removed.

**Comment by ScipioWright** [comment]
Converting to draft until I have the time to fix the merge conflicts and deal with the comments

**Review by Berserker66 — APPROVED** [review]
